pkg/apply/prune/prune.go

```diff
@@ -77,55 +73,57 @@ type Options struct {
 }
 
-// Prune deletes the set of passed pruneObjs.
+// Prune deletes the set of passed pruneObjs. A prune skip/failure is
+// captured in the TaskContext, so we do not lose track of these
+// objects from the inventory. The passed prune filters are used to
+// determine if permission exists to delete the object. An example
+// of a prune filter is PreventDeleteFilter, which checks if an
+// annotation exists on the object to ensure the objects is not
+// deleted (e.g. a PersistentVolume that we do no want to
+// automatically prune/delete).
 //
 // Parameters:
-//   localInv - locally read inventory object
 //   pruneObjs - objects to prune (delete)
-//   currentUIDs - UIDs for successfully applied objects
+//   pruneFilters - list of filters for deletion permission
 //   taskContext - task for apply/prune
-func (po *PruneOptions) Prune(localInv inventory.InventoryInfo,
-	pruneObjs []*unstructured.Unstructured,
-	currentUIDs sets.String,
+//   o - options for dry-run
+func (po *PruneOptions) Prune(pruneObjs []*unstructured.Unstructured,
+	pruneFilters []filter.ValidationFilter,
 	taskContext *taskrunner.TaskContext,
 	o Options) error {
-	// Validate parameters
-	if localInv == nil {
-		return fmt.Errorf("the local inventory object can't be nil")
-	}
-	// Sort the resources in reverse order using the same rules as is
-	// used for apply.
 	eventFactory := CreateEventFactory(po.Destroy)
+	// Iterate through objects to prune (delete). If an object is not pruned
+	// and we need to keep it in the inventory, we must capture the prune failure.
 	for _, pruneObj := range pruneObjs {
 		pruneID := object.UnstructuredToObjMeta(pruneObj)
 		klog.V(5).Infof("attempting prune: %s", pruneID)
-		// Do not prune objects that are in set of currently applied objects.
-		uid := string(pruneObj.GetUID())
-		if currentUIDs.Has(uid) {
-			klog.V(5).Infof("prune object in current apply; do not prune: %s", uid)
-			continue
-		}
-		// Handle lifecycle directive preventing deletion.
-		if !canPrune(localInv, pruneObj, o.InventoryPolicy, uid) {
-			klog.V(4).Infof("skip prune for lifecycle directive %s", pruneID)
-			taskContext.EventChannel() <- eventFactory.CreateSkippedEvent(pruneObj)
-			taskContext.CapturePruneFailure(pruneID)
-			continue
+		// Check filters to see if we're prevented from pruning/deleting object.
+		var filtered bool
+		var err error
+		for _, filter := range pruneFilters {
+			klog.V(6).Infof("prune filter %s: %s", filter.Name(), pruneID)
+			filtered, err = filter.Filter(pruneObj)
+			if err != nil {
+				if klog.V(5).Enabled() {
+					klog.Errorf("error during %s, (%s): %s", filter.Name(), pruneID, err)
+				}
+				taskContext.EventChannel() <- eventFactory.CreateFailedEvent(pruneID, err)
+				taskContext.CapturePruneFailure(pruneID)
+				break
+			}
+			if filtered {
+				klog.V(4).Infof("prune filtered by %s: %s", filter.Name(), pruneID)
+				taskContext.EventChannel() <- eventFactory.CreateSkippedEvent(pruneObj)
```

Comment:
It seems like we need some way of including the reason why a resource was skipped here. It would be pretty useful to know if it is due to a lifecycle directive (and in that case, which one), because it is a namespace in use by one or more applied objects, or something different.

Reply:
We will add a `Reason` function in the `ValidationFilter` interface to replace the `Name` function. Also, we will update the `CreateSkippedEvent` function signature to take the reason as a string. We should probably also update the `PruneEvent` struct to add a `Reason` field, but we should probably do that in another PR. Please let me know if this direction is what you had in mind.

Reply:
Since this can be a large change, and since this PR is already large, we should probably put this refactor work off to another PR.